### PR TITLE
Support for algebraic variables and templates update

### PR DESCRIPTION
**Pull request by zanellia**
With this pull request (which replaces https://github.com/acados/acados/pull/360) we add support for algebraic variables in linear least-squares costs:

|| Vx * x + Vu * u + Vz * z ||^2_W.

Additional changes:

1. Updated templating interface to support DAEs and parameters in the dynamics.
2. Some more work on generation of S-Functions from the templating interface.
3. Support for terminal algebraic constraints
4. Support for convex-composite constraints from templating interface

@FreyJo already had a look at most of the code related to algebraic variables, maybe @giaf can also have a quick look? :)

PS: Travis keeps failing on macOS. Any Mac user could assist me for 10 minutes at some point? Maybe @schoelst? :)

### CODE REVIEW

File: examples/c/simple_dae_model/export_simple_dae_model.m [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    %% CasADi
+    import casadi.*
+    if CasadiMeta.version()=='3.4.0'
+        % casadi 3.4
+        casadi_opts = struct('mex', false, 'casadi_int', 'int', 'casadi_real', 'double');
+    else
+        % old casadi versions
+        error('Please download and install Casadi 3.4.0 to ensure compatibility with acados')
+    end
+    model_name_prefix = 'inv_pendulum';
```
Comment (FreyJo) [inline examples/c/simple_dae_model/export_simple_dae_model.m]: I made a PR to your PR, such that the model is generated using the functions in `external_function_generation`.

File: interfaces/acados_template/acados_template/generate_c_code_implicit_ode.py [hunk 2]
```diff
@@ -95,11 +95,18 @@ def generate_c_code_implicit_ode( model, opts ):
...
-        fun_name = model_name + '_impl_dae_jac_x_xdot_u'
-        impl_dae_jac_x_xdot_u = Function(fun_name, [x, xdot, u, z, p], [jac_x, jac_xdot, jac_u, jac_z])
+        # fun_name = model_name + '_impl_dae_fun_jac_x_xdot_z'
+        # impl_dae_fun_jac_x_xdot = Function(fun_name, [x, xdot, u, z, p], [f_impl, jac_x, jac_xdot, jac_z])
+
+        # fun_name = model_name + '_impl_dae_jac_x_xdot_u'
+        # impl_dae_jac_x_xdot_u = Function(fun_name, [x, xdot, u, z, p], [jac_x, jac_xdot, jac_u, jac_z])
         
         fun_name = model_name + '_impl_dae_fun_jac_x_xdot_u_z'
         impl_dae_fun_jac_x_xdot_u = Function(fun_name, [x, xdot, u, z, p], [f_impl, jac_x, jac_xdot, jac_u])
```
Comment (FreyJo) [inline interfaces/acados_template/acados_template/generate_c_code_implicit_ode.py]: I guess `_z` should be removed from the name, since `jac_z` is not part of output

**Comment by zanellia** [comment]
Thanks for the review @FreyJo! I will implement your comments next week ;)

File: examples/c/inverted_pendulum_model/generate_casadi_functions.m [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+%   You should have received a copy of the GNU Lesser General Public
+%   License along with acados; if not, write to the Free Software Foundation,
+%   Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+%
+
+clc;
+clear all;
+close all;
+
+addpath('/home/andrea/casadi-linux-octave-v3.4.0')
```
[the remark this answers is not in the document]
Comment (zanellia) [reply examples/c/inverted_pendulum_model/generate_casadi_functions.m]: I have removed this, thanks ;) I still leave the script in this folder as @giaf moved the script in interfaces/matlab/sim to a different location and slightly adapted it. I would be happy to remove it and refer to the common script as soon as we stop updating things :)

File: examples/c/CMakeLists.txt [hunk 4]
```diff
@@ -219,14 +219,31 @@ add_executable(sim_gnsf_crane sim_gnsf_crane.c ${GNSF_CRANE_SRC})
...
-add_test(pendulum_scqp pendulum_scqp)
-
+# # -------------------- dae_example
+# add_executable(dae_example dae_example.c
+#     inverted_pendulum_model/impl_ode_fun_pendulum_dae.c
+#     inverted_pendulum_model/impl_ode_jac_x_xdot_u_z_pendulum_dae.c
+#     inverted_pendulum_model/impl_ode_fun_jac_x_xdot_z_pendulum_dae.c)
+    
+# target_link_libraries(dae_example acados)
+# add_test(dae_example dae_example)
```
[the remark this answers is not in the document]
Comment (zanellia) [reply examples/c/CMakeLists.txt]: fixed.

File: interfaces/acados_template/acados_template/generate_c_code_implicit_ode.py (hunk 2 above)
Comment (zanellia) [reply interfaces/acados_template/acados_template/generate_c_code_implicit_ode.py]: fixed with e2421d.

**Comment by zanellia** [comment]
OK, I eventually added a wrapper to the ctypes calls for the Python interface. No big changes with respect to e2421dd.

File: acados/ocp_nlp/ocp_nlp_sqp.c [hunk 5]
```diff
@@ -810,14 +811,15 @@ static void linearize_update_qp_matrices(void *config_, ocp_nlp_dims *dims, ocp_
...
     {
-        // cost
-        config->cost[i]->update_qp_matrices(config->cost[i], dims->cost[i], nlp_in->cost[i],
-                                            opts->cost[i], mem->cost[i], work->cost[i]);
         // dynamics
         if (i < N)
             config->dynamics[i]->update_qp_matrices(config->dynamics[i], dims->dynamics[i],
                                                 nlp_in->dynamics[i], opts->dynamics[i],
                                                 mem->dynamics[i], work->dynamics[i]);
+        // cost
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_sqp.c]: Why did you move the cost update after the dynamics? For the way it is coded, it should be before: `cost` initializes the hessian of the QP, and dynamics makes an update to it in case of exact hessian

File: acados/ocp_nlp/ocp_nlp_cost_common.h [hunk 6]
```diff
@@ -59,6 +59,8 @@ typedef struct
     int (*memory_calculate_size)(void *config, void *dims, void *opts);
     struct blasfeo_dvec *(*memory_get_grad_ptr)(void *memory);
     void (*memory_set_ux_ptr)(struct blasfeo_dvec *ux, void *memory);
+    void (*memory_set_z_ptr)(struct blasfeo_dvec *z, void *memory);
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_cost_common.h]: The name `z` is already taken for the linear cost of the slack variables `s` (the quadratic cost is `Z`).

As it is coded now, this is unclear if `memory_set_z_ptr` refers to the cost `z` or the algebraic variable `z`.
The solution is to rename one of the two.
Some ideas: `z_alg` for the algebraic variable, or conversely `Zs` and `zs` for the costs of the slack variables.
Other ideas?

File: acados/ocp_nlp/ocp_nlp_cost_common.h [hunk 7]
```diff
@@ -59,6 +59,8 @@ typedef struct
...
+    void (*memory_set_dzdxu_tran_ptr)(struct blasfeo_dmat *dzdxu, void *memory);
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_cost_common.h]: Why (dz)/(dxu), if it is used as (dz)/(dux) ?

File: acados/ocp_nlp/ocp_nlp_cost_ls.c [hunk 8]
```diff
@@ -607,22 +661,51 @@ void ocp_nlp_cost_ls_update_qp_matrices(void *config_, void *dims_,
...
     int nu = dims->nu;
+    int nz = dims->nz;
     int ny = dims->ny;
     int ns = dims->ns;
 
     // initialize hessian of lagrangian with hessian of cost
     blasfeo_dgecp(nu + nx, nu + nx, &memory->hess, 0, 0, memory->RSQrq, 0, 0);
 
+    // copy Cyt into Cyt_tilde
+    blasfeo_dgecp(nu + nx, ny, &model->Cyt, 0, 0, &work->Cyt_tilde, 0, 0);
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_cost_ls.c]: Avoid unnecessary copy for the most common case of `nz=0`, and keep the old code with an `if ... else ...`

File: acados/ocp_nlp/ocp_nlp_cost_ls.c [hunk 9]
```diff
@@ -607,22 +661,51 @@ void ocp_nlp_cost_ls_update_qp_matrices(void *config_, void *dims_,
...
+    // copy y_ref into y_ref_tilde
+    blasfeo_dveccp(ny, &model->y_ref, 0, &work->y_ref_tilde, 0);
+    if (nz > 0) { // eliminate algebraic variables and update Cyt and y_ref
+        // swap dzdu and dzdx
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_cost_ls.c]: Put in the correct format `dzdux_tran` directly when getting the sensitivities from the integrator's `S_algebraic`: there is no use of `dzdxu`, right?

File: acados/ocp_nlp/ocp_nlp_cost_ls.c [hunk 10]
```diff
@@ -607,22 +661,51 @@ void ocp_nlp_cost_ls_update_qp_matrices(void *config_, void *dims_,
...
+        blasfeo_dgecp(nx, ny, memory->dzdxu_tran, 0, 0,  &work->dzdux_tran, nu, 0);
+        blasfeo_dgecp(nu, ny, memory->dzdxu_tran, nx, 0, &work->dzdux_tran, 0, 0);
+        // update Cyt: Cyt_tilde = Cyt + dzdux_tran*Vz^T
+        blasfeo_dgemm_nt(nu + nx, ny, nz, 1.0, &work->dzdux_tran, 0, 0,
+                &model->Vz, 0, 0, 1.0, &work->Cyt_tilde, 0, 0, &work->Cyt_tilde, 0, 0);
+
+        // update y_ref: y_ref_tilde = y_ref + Vz*x + Vz*u - Vz*z
+        blasfeo_dveccp(nz, &memory->z, 0, &work->tmp_nz, 0);
```
Comment (giaf) [inline acados/ocp_nlp/ocp_nlp_cost_ls.c]: No need to copy: this can be taken care from the 3rd argument of the next `blasfeo_dgemv_t`

File: acados/ocp_nlp/ocp_nlp_sqp.c (hunk 5 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_sqp.c]: The problem is that when you want to have algebraic variables in the cost, `dynamics` necessarily needs to be called *before* `cost`, due to the fact that the sensitivities dz/dux need to be computed (it should actually be the same with nonlinear least-squares costs). I get your point though, this way the exact hessian update would not be correct. I can think about how to fix this, do you have a solution already?

File: acados/ocp_nlp/ocp_nlp_cost_common.h (hunk 6 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_cost_common.h]: addressed here: 2504e0f (z -> z_alg, for the time being)

File: acados/ocp_nlp/ocp_nlp_sqp.c (hunk 5 above)
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_sqp.c]: I guess we need a zero-initialization of the Hessian ( H=0 ) before anything, and just update ( H += ... ) in cost/dynamics/constraints, so order does not matter, and you can freely swap dynamics before cost

File: acados/ocp_nlp/ocp_nlp_cost_common.h (hunk 6 above)
Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_cost_common.h]: great thanks :)

File: acados/ocp_nlp/ocp_nlp_cost_ls.c (hunk 9 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: right, addressed in 940fc22.

File: acados/ocp_nlp/ocp_nlp_cost_ls.c (hunk 8 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: addressed in 9319a5.

File: acados/ocp_nlp/ocp_nlp_cost_ls.c (hunk 10 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_cost_ls.c]: addressed here f4181ec

File: acados/ocp_nlp/ocp_nlp_sqp.c (hunk 5 above)
Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_sqp.c]: I have made an attempt to do this here: 7bd9153. Please have a look, I do not know how to test this thoroughly!

Comment (giaf) [reply acados/ocp_nlp/ocp_nlp_sqp.c]: It looks fine, but it should be applied also to `sqp_rti` and `cost_external`

Comment (zanellia) [reply acados/ocp_nlp/ocp_nlp_sqp.c]: addressed here 7bd9153 - b30f85a. Again, I have no idea about how to test this properly, please have a look/tell me how to test it :)

**Comment by giaf** [comment]
Just put here https://github.com/acados/acados/blob/b30f85acd2314cba098e99ed5b4c0671251f654b/acados/ocp_nlp/ocp_nlp_cost_external.c#L511 `work->tmp_nu_nx` and make the sum `memory->RSQrq += work->tmp_nu_nx` and we are good ;)

A smaller note: the name `tmp_nu_nx` suggests something of size (nu)x(nx), so I would go for (nv)x(nv), where nv = nu+nx, but this is minor and I can change myself later when working on this.

If you can quickly fix the cost_external, I would merge immediately and start working on Hessian myself.

**Comment by zanellia** [comment]
> Just put here
> 
> https://github.com/acados/acados/blob/b30f85acd2314cba098e99ed5b4c0671251f654b/acados/ocp_nlp/ocp_nlp_cost_external.c#L511
> 
> `work->tmp_nu_nx` and make the sum `memory->RSQrq += work->tmp_nu_nx` after the function call and we are good ;)
> A smaller note: the name `tmp_nu_nx` suggests something of size (nu)x(nx), so I would go for (nv)x(nv), where nv = nu+nx, but this is minor and I can change myself later when working on this.
> 
> If you can quickly fix the cost_external, I would merge immediately and start working on Hessian myself.

sorry, that was stupid! That's why we need proper tests ;)
